cloudinit/sources/helpers/azure.py

```diff
@@ -525,19 +533,41 @@ def send_ready_signal(self) -> None:
 
         LOG.info('Reported ready to Azure fabric.')
 
+    @azure_ds_telemetry_reporter
+    def send_failure_signal(self, description: str) -> None:
+        document = self.build_report(
+            incarnation=self._goal_state.incarnation,
+            container_id=self._goal_state.container_id,
+            instance_id=self._goal_state.instance_id,
+            status=self.PROVISIONING_NOT_READY_STATUS,
+            substatus=self.PROVISIONING_FAILURE_SUBSTATUS,
+            description=description)
+        LOG.warning('Reporting failure to Azure fabric.')
+        try:
+            self._post_health_report(document=document)
+        except Exception as e:
+            msg = "exception while reporting failure: %s" % e
+            LOG.error(msg)
+            report_diagnostic_event(msg, logger_func=LOG.error)
+            raise
+
+        LOG.warning('Reported failure to Azure fabric.')
+
     def build_report(
             self, incarnation: str, container_id: str, instance_id: str,
             status: str, substatus=None, description=None) -> str:
         health_detail = ''
         if substatus is not None:
             health_detail = self.HEALTH_DETAIL_SUBSECTION_XML_TEMPLATE.format(
-                health_substatus=substatus, health_description=description)
+                health_substatus=escape(substatus),
+                health_description=escape(
+                    description[:self.HEALTH_REPORT_DESCRIPTION_TRIM_LEN]))
```

Comment:
The Azure platform has a 4096 charlen limit for the description XML field within the XML health doc. For now in this PR, the description message is solely controlled by us. In the future, there may be plans to expose parts of the exception message to the user, so trimming is necessary.

Trimming before escaping also prevents invalid XML. The short trim length (512) vs the platform limit len (4096) also prevents the situation where the trimmed XML blows past length limits after escaping (since XML escaping lengthens a string).

Trimming is also necessary for user experience considerations. Wouldn't want an overly long error message.